apiStatus annotation, user-land compilation warning/errors

eed3si9n opened the pull request:
This implements `apiStatus` annotation as a generic form of `deprecated` annotation. `apiStatus` takes `category` and `defaultAction` parameters, corresponding to configurable warning's category and action.

One of the usage is to trigger compiler error from the library when a method is invoked to display migration message. Another usage would be to denote bincompat status of the API as warning.

This is a resend of #7790 based on the configurable warnings.
Ref #8373 / https://twitter.com/not_xuwei_k/status/1240354073297268737
Ref https://github.com/scala/scala/pull/7528

### From Scaladoc

An annotation to denote the API status, like `scala.deprecated` but more powerful.
While `@deprecated` is only able to discourage someone from using the some API,
`@apiStatus` can be more nuanced about the state (for instance `Category.ApiMayChange`),
and choose the default compile-time actions (`Action.Error`, `Action.Warning`, etc).

In other words, this gives library authors the lever to trigger compilation warning or
compilation errors! Here's an example of displaying a migration message:

```scala
import scala.annotation.apiStatus, apiStatus._
@apiStatus(
  "method <<= is removed; use := syntax instead",
  category = Category.ForRemoval,
  since = "foo-lib 1.0",
  defaultAction = Action.Error,
)
def <<=() = ???
```

The compilation will fail and display the migration message if the method is called:

```scala
example.scala:26: error: method <<= is removed; use := syntax instead (foo-lib 1.0)
  <<=()
  ^
```

Here's another example of displaying a warning:

```scala
import scala.annotation.apiStatus, apiStatus._
@apiStatus(
  "should DSL is incubating, and future compatibility is not guaranteed",
  category = Category.ApiMayChange,
  since = "foo-lib 1.0",
)
implicit class ShouldDSL(s: String) {
  def should(o: String): Unit = ()
}
```

The compiler will emit warnings:

```scala
example.scala:28: warning: should DSL is incubating, and future compatibility is not guaranteed (foo-lib 1.0)
  "bar" should "something"
  ^
```

Using `-Wconf:cat=api-may-change&origin=foo\..*:silent` option, the user of the library can opt out of the `api-may-change` warnings afterwards.

### Defining a custom status annotation
 
Instead of directly using `@apiStatus` we can create a specific status annotation by extending `apiStatus`. However, due to the information available to the compiler, the default values for `message`, `category`, or `defaultAction` can be specified through the corresponding meta-annotations.
 
```scala
import scala.annotation.{ apiStatus, apiStatusCategory, apiStatusDefaultAction }
import scala.annotation.meta._

@apiStatusCategory(apiStatus.Category.ApiMayChange)
@apiStatusDefaultAction(apiStatus.Action.Warning)
@companionClass @companionMethod
final class apiMayChange(
  message: String,
  since: String = "",
) extends apiStatus(message, since = since)
```

 This can be used as follows:

```scala
@apiMayChange("should DSL is incubating, and future compatibility is not guaranteed")
implicit class ShouldDSL(s: String) {
  def should(o: String): Unit = ()
}
 ```

sjrd commented:
What's wrong with `@compileTimeOnly`?

som-snytt commented on src/library/scala/deprecated.scala:
```diff
@@ -55,7 +55,21 @@ import scala.annotation.meta._
+@deprecatedInheritance("Scheduled for being sealed in 2.14", "2.13.0")
 class deprecated(message: String = "", since: String = "") extends scala.annotation.StaticAnnotation
+
+/** An annotation that designates that an annottee should no longer be used.
```
We must correct the spelling of the neologism. `annotatee` (what is annotated by an annotator) or `annotee` (anglicization).

Other good words from checking the OED: annonary (provisions), annotine (year-old), annosity (fullness of years), annoyously.

som-snytt replied:
Sample contemporary usage of annonary: https://sandus.org/central-peoples-government/office-of-the-sogmo/nobility-honours/fraternal-annonary-order-of-the-wooden-bowl/

eed3si9n replied:
The Annotated. An undercover grad student and a mole in Lightbend attempt to identify each other while combing through scala-internals threads for the originalist interpretation of Scala Language Spec.

som-snytt commented:
In line with "deprecation" as a curse, `@anathema`.

eed3si9n commented:
@sjrd 

> What's wrong with `@compileTimeOnly`?

I guess the name is wrong. If I'm removing `def <<=` I want to say that it's an error, typer phase or otherwise.

som-snytt commented:
Someone claimed previously that `@deprecated` in Scala always means "forRemoval" (which I think is patently false, witness the `gazillion deprecations since 2.10` messages, except now it says `100000000` instead of writing out gazillion). I may have PRd "forRemoval"?

Maybe it's coronavirus, sheltering in place, the general need to declutter my life, but now I want no warnings from the compiler, but better integration with tooling for diagnostics. I don't even want deprecations from the compiler, but from a linter, which can decide what is an error and what can be ignored.

It's also true that `compileTimeOnly` is still a compile-time dependency. If the intention is to error because `forRemoval`, then it doesn't suffice; the element could be used but eliminated by a macro or `elidable`. If the intention is just don't call me, because the implementation is `???`, then an alias of `compileTimeOnly` would work.

lrytz commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -26,7 +26,12 @@ trait Warnings {
-  private val WconfDefault = List("cat=deprecation:ws", "cat=feature:ws", "cat=optimizer:ws")
+  private val WconfDefault = List(
+    "cat=deprecation:ws",
+    "cat=deprecation-error:error",
```
```suggestion
    "cat=deprecation-error:e",
```

eed3si9n replied:
I used the longer option intentionally for the purpose of readability. I think I came across the idea here https://kvz.io/bash-best-practices.html a while back:

> Use long options (logger --priority vs logger -p). If you're on cli, abbreviations make sense for efficiency. but when you're writing reusable scripts a few extra keystrokes will pay off in readability and avoid ventures into man pages in the future by you or your collaborators.

lrytz reviewed (commented):
To me, https://github.com/scala/scala/pull/7790 always felt like addressing a mixed bag of concerns and I didn't manage to grasp "what's the goal", that's why I was concerned.

Now this PR is different, I can understand what it does and I can see how some projects would want to use it, so I'm fine with adding it.

🎨 how about adding a parameter to `deprecated` instead? `@deprecated(message = "...", since = "...", error = true)`

dwijnand commented:
`@compileTimeOnly("use := syntax instead")` looks good enough, without adding more copies.  Perhaps all this PR needs to do is add this example to its Scaladoc?

dwijnand commented on test/files/run/settings-parse.check:
```diff
@@ -12,55 +12,55 @@
 3) List(-cp, , -deprecation) ==> Settings {
   -deprecation = true
-  -Wconf = List(cat=deprecation:w, cat=deprecation:ws, cat=feature:ws, cat=optimizer:ws)
+  -Wconf = List(cat=deprecation:w, cat=deprecation:ws, cat=deprecation-error:error, cat=feature:ws, cat=optimizer:ws)
```
~Well `Wconf.clearSetByUser()` clearly isn't working :-/~. Oh, nvm, these are set because of the `-deprecation`.

eed3si9n commented:
@lrytz 

> 🎨 how about adding a parameter to deprecated instead? `@deprecated(message = "...", since = "...", error = true)`

I thought `DeprecationError` should be a category so if someone wants to ignore it (and deal with the consequences) they can using -Wconf.

lrytz commented:
> `DeprecationError` should be a category

I agree, but that could be done with the additional parameter too

eed3si9n commented:
> I agree, but that could be done with the additional parameter too

That's true. I can make `Reporting#deprecationWarning` do the deprecation error depending on the parameter.

lrytz commented on src/library/scala/annotation/apiStatus.scala:
```diff
@@ -0,0 +1,53 @@
+@getter @setter @beanGetter @beanSetter @companionClass @companionMethod
+class apiStatus(
+  message: String,
+  category: String,
```
maybe this should be just called "status", 1 to avoid introducing yet another term to the problem space here, and 2 to avoid overlap with `-Wconf`'s "category"

lrytz replied:
Ah, you're actually integrating with the Wconf categories! Then that makes sense, I'll think about this more.

lrytz commented on src/library/scala/annotation/apiStatus.scala:
```diff
@@ -0,0 +1,53 @@
+    final val Error = "error"
+    final val Warning = "warning"
+    final val WarningSummary = "warning-summary"
+    final val WarningVerbose = "warning-verbose"
```
Probably we should not expose the "verbose" options here?

eed3si9n replied:
👍

eed3si9n replied:
Fixed in https://github.com/scala/scala/pull/8820/commits/9f6413c4e9fcef749bc38040e250706df9b82d21. Also added Scaladoc in there.

eed3si9n committed:
apiStatus annotation

This implements `apiStatus` annotation as a generic form of `deprecated` annotation. While deprecated is only able to discourage someone from using the some API, `apiStatus` can be more nuanced about the state (for instance Category.ApiMayChange), and choose the default compile-time actions (Action.Error, Action.Warning, etc). In other words, this gives library authors the lever to trigger compilation warning or
compilation errors!

One of the usage is to trigger compiler error from the library when a method is invoked to display migration message. Another usage would be to denote bincompat status of the API as warning.

This is a resend of #7790 based on the configurable warnings.
Ref #8373 / https://twitter.com/not_xuwei_k/status/1240354073297268737

hepin1989 commented:
I think it would be better annotate the apiMayChange directly not the ApiStatus

eed3si9n commented:
@hepin1989 

> I think it would be better annotate the apiMayChange directly not the ApiStatus

Different library authors want different ways of emitting compiler errors and warnings. Rather than adding each piecemeal, `@apiStatus` provides a toolbox so they can add different categories themselves.

For example, some may opt to display [ApiMayChange](https://doc.akka.io/docs/akka/current/common/may-change.html) as a warning, while others may want to keep it silent by default, and let user opt into it as a warning. Same could be said of other more nuanced notifications such as semantics change, performance regression, or non-deterministic output.

nafg commented:
I feel like the name apiStatus refers to particular use case rather than describing what it's the name of. Can I suggest something like compilationProblem instead? ("Problem" seems to be the term IDEs use to lump errors and warnings together, so something that can be either an error or warning is called a problem.)

eed3si9n commented:
@nafg `@compileTimeAction` might be a better procedural name, since _compilation problem_ sounds like you're talking about a bug in compiler.

`@apiStatus` is meant to be a metadata of the API, like it's intended for internal-use. The fact that it's internal is not a problem per se.

eed3si9n committed:
Implement extension mechanism for apiStatus

Library authors can extend apiStatus attribute to provide a custom status annotation.

Due to the information available during compilation, the default values for `message`, `category`, or `defaultAction` are specified through the corresponding meta-annotations.

eed3si9n commented:
Added a few meta-annotations so we can define custom annotations:

```scala
import scala.annotation.{ apiStatus, apiStatusCategory, apiStatusDefaultAction }
import scala.annotation.meta._
 
@apiStatusCategory("api-may-change")
@apiStatusDefaultAction(apiStatus.Action.Warning)
@companionClass @companionMethod
final class apiMayChange(
  message: String,
  since: String = "",
) extends apiStatus(message, since = since)
```

This would simplify the tagging, comparable to `deprecated` etc:

```scala
@apiMayChange("can DSL is incubating, and future compatibility is not guaranteed")
implicit class CanDSL(s: String) {
  def can(o: String): Unit = ()
}
```

DryPilgrim commented:
![image](https://user-images.githubusercontent.com/44194595/89123632-553e8400-d503-11ea-85c9-db957dd0d31f.png)

SethTisue commented:
Perhaps this could use a round of team discussion.

My starting position is something like:

That I'm far from convinced that we need anything like either Eugene's original proposal or the followup one with meta-annotations. It just seems unnecessarily elaborate to me and as a result I'm skeptical that library authors would actually learn and use it. The more minimal alternatives suggested by Dale and Lukas appeal to me more.

eed3si9n commented:
Note that my original PR #7790 for `deprecationError` I sent in March 2019 was really simple - https://github.com/scala/scala/pull/7790/commits/f890cbf2c3d22c417a8a3c496cb79dfd0d4ea780

It gradually evolved into this as I was addressing reviewers' suggestions. Before we repeat another go-around, could we decide if Scala 2.x is at all interested in having a user-land compilation errors or user-land warnings? And come up with criteria of what would make the solution merge-able?

dwijnand commented:
I don't think we have the bandwidth with the time remaining, but let's try to discuss this with you for the next release.

lrytz commented:
> Before we repeat another go-around, could we decide if Scala 2.x is at all interested in having a user-land compilation errors or user-land warnings? And come up with criteria of what would make the solution merge-able?

I very much agree. For the first question, I don't know myself.

If the answer is yes, I also find it difficult to know what exactly is needed / will be used in reality by the various library authors. For this reason, a very general and configurable solution like it's implemented here makes sense to me (the details should be discussed of course). The simpler proposals in the past always caused a lot of questions in both directions: why not just use existing feature X? Why doesn't it support use case Y?

lrytz commented:
@eed3si9n I played a bit around with new meta-annotations for defaults which would allow changing

```scala
@apiStatusCategory(apiStatus.Category.ApiMayChange)
@apiStatusDefaultAction(apiStatus.Action.Warning)
@companionClass @companionMethod
final class apiMayChange(
  message: String,
  since: String = "",
) extends apiStatus(message, since = since)
```

to

```scala
@companionClass @companionMethod
final class apiMayChange(
  message: String,
  since: String = "",
) extends apiStatus(
  message,
  category = apiStatus.Category.ApiMayChange,
  since = since,
  defaultAction = apiStatus.Action.Warning)
```

See the last commit message in https://github.com/scala/scala/compare/2.13.x...lrytz:constAnnDefaults

eed3si9n commented:
> ```scala
> 
> @companionClass @companionMethod
> 
> final class apiMayChange(
> 
>   message: String,
> 
>   since: String = "",
> 
> ) extends apiStatus(
> 
>   message,
> 
>   category = apiStatus.Category.ApiMayChange,
> 
>   since = since,
> 
>   defaultAction = apiStatus.Action.Warning)
> 
> ```
> 
> See the last commit message in https://github.com/scala/scala/compare/2.13.x...lrytz:constAnnDefaults

That's great. Much more straightforward.

SethTisue commented:
relevant: Scala 3 appears set to add `@experimental`: https://github.com/lampepfl/dotty/pull/12102

SethTisue commented:
It remains unclear whether this is ever going to progress. Closing to get it out of the PR queue, at least for the time being.
